server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -696,6 +696,32 @@ static String getCommitCheckpointFileName(long generation) {
         return TRANSLOG_FILE_PREFIX + generation + CHECKPOINT_SUFFIX;
     }
 
+    /**
+     * Trims translog for terms of files below <code>belowTerm</code> and seq# above <code>aboveSeqNo</code>.
+     * Effectively it moves max visible seq# {@link Checkpoint#trimmedAboveSeqNo} therefore {@link TranslogSnapshot} skips those operations.
+     */
+    public void trim(long belowTerm, long aboveSeqNo) throws IOException {
+        assert aboveSeqNo > 0 : "aboveSeqNo has to be positive number";
+        try (ReleasableLock lock = writeLock.acquire()) {
+
+            // update all existed ones (if it is necessary) as checkpoint and reader are immutable
+            final List<TranslogReader> newReaders = new ArrayList<>(readers.size());
+            try {
+                for (TranslogReader reader : readers) {
+                    final TranslogReader newReader = reader.closeIntoTrimmedReader(belowTerm, aboveSeqNo, getChannelFactory());
```

Comment:
I think you need to check whether the reader actually needs to be closed. See my comments on `closeIntoTrimmedReader`.